docs/documentation/upgrading/topics/changes/changes-25_0_0.adoc

```diff
@@ -80,6 +80,91 @@ In order to dynamically resolve it from request headers, you need to set the `ho
+
+== Migrating user sessions during the upgrade
+
+When upgrading from {project_name} 24 or earlier, admins can choose to migrate existing online user and client sessions to persistent sessions.
+For this to work, those existing sessions need to be stored in either a remote {jdgserver_name} or in a database configured as JDBC persistence for {project_name}'s embedded cache.
+Migrating in-memory sessions for {project_name} 24 is not supported as all {project_name} instances need to be shut down before the upgrade due to a major version upgrade of the embedded cache.
+
+[WARNING]
+====
+The migration of user sessions only works when the persistent user sessions is enabled when upgrade to {project_name} 25.
+If you chose to upgrade to 25 without enabling persistent user sessions, there is currently no possibility to trigger the migration of existing sessions at a later point in time.
+When you enable it later, only newly created user sessions will be persisted.
+====
+
+To migrate the user sessions during an upgrade of {project_name}, perform the following steps:
+
+. Stop all running old instances of {project_name}.
+
+. Create backups:
++
+--
+* Create a backup {project_name}'s database.
+* If JDBC persistence is used, create a backup of that database if you want to be able to retry the migration of the sessions.
+* If an external {jdgserver_name} is used, create a backup of its data if you want to be able to retry the migration of the sessions.
+--
+. Start the new instances {project_name} with the persistent user sessions feature enabled.
++
+The first starting node will:
++
+--
+. Migrate the database to the schema version 25.
+. Copy all session information from either the remote {jdgserver_name} or the JDBC persistence configured for {project_name}'s embedded cache to the database of {project_name}.
++
+The data will be stored in the tables `offline_user_session` and `online_user_session` with `offline_flag` set to `false`.
+
+. Clear the caches.
++
+This includes clearing the caches of the external {jdgserver_name} if one is used, and clearing the JDBC persistence if one is used.
+--
+. Update the cache configuration XML of {project_name} for caches `sessions` and `clientSessions`:
++
+--
+* If JDBC persistence is used, remove the configuration for JDBC persistence.
+* If the remote {jdgserver_name} has been used in a single-site setup solely for keeping user sessions across {project_name} restarts, remove the remote {jdgserver_name} configuration for those caches.
+* If the remote {jdgserver_name} is used in a multi-site setup, update the cache config as outlined in https://www.keycloak.org/high-availability/deploy-infinispan-kubernetes-crossdc[Deploy Infinispan for HA with the Infinispan Operator] {section} with the number of owners and the maximum memory size to reduce the memory usage.
```

Comment:
I'll move this to a "tip" or "info" panel with the following semantics: you can reduce the resource consumption by the external Infinispan by configuring the number of entries in memory.

Reply:
done in next commit